### PR TITLE
Add doc about cpp wasm extension development

### PR DESCRIPTION
**Pull request by bianpengyuan**
Signed-off-by: Pengyuan Bian <bianpengyuan@google.com>

### CODE REVIEW

**Comment by bianpengyuan** [comment]
cc @gargnupur please take a look thanks!

File: doc/development-setup.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Set up Develop Environment
```
Comment (gargnupur) [inline doc/development-setup.md:1]: should envoy development link be added too? Once users have that.. all of the things will follow?

File: doc/write-a-wasm-extension-with-cpp.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+ addResponseHeader("X-Wasm-custom", "foo");
+ return FilterHeadersStatus::Continue;
+}
+```
+
+After adding the method, build the extension again with the bazel command mentioned in step 2 and the `example.wasm` file should be updated now with the new logic.
+
+## Step 4: Write integration test
+
+To enable quick iteration and deploy your extension with confidence, it is **highly** recommended to write integration tests with your extension and the same Envoy binary Istio proxy runs. To achieve this, the same Golang integration test framework for Telemetry v2 filter could be used, which at high level does the follows:
```
Comment (gargnupur) [inline doc/write-a-wasm-extension-with-cpp.md:73]: add a link to the telemetry v2 integration framework?

File: doc/write-a-wasm-extension-with-cpp.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+An unit test example could be found under the [basic auth plugin](../extensions/basic_auth/plugin_test.cc), where several callbacks are mocked, such as `getBuffer` is mocked to return desired configuration, `getHeaderMapValue` is mocked to return wanted header pair. The whole test needs to be wrapped with `null_plugin` namespace. To learn more about how to write unit tests for the extension, please refer to this [guide](./write-unit-test.md).
+
+## Step 6: Push and deploy the extension
+
+After the extension has been verified and tested, it is time to deploy the extension with Istio! Assume you have already deployed a [httpbin example app](https://github.com/istio/istio/tree/master/samples/httpbin) in your cluster, and the extension has been pushed to a blob serving service, applying the following EnvoyFilter to inject the example extension:
+
+Apply the following [EnvoyFilter](../example/config/storage-cluster.yaml) to register the blob service cluster for Extension downloading. Here we use Google Cloud Storage as an example:
```
Comment (gargnupur) [inline doc/write-a-wasm-extension-with-cpp.md:91]: This is cool.. I hadn't tried this one!

File: doc/write-a-wasm-extension-with-cpp.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+$ curl -I 34.123.215.139
+HTTP/1.1 200 OK
+server: istio-envoy
+...
+x-wasm-custom: foo
+```
+
+## Step 7: Maintain your extension along with Istio releases
+
+After writing, testing, and deploying the extension successfully, it is time to make sure that your extension could evolve along with Istio. Currently it is recommended to build and deploy your extension following Istio releases. Specifically:
```
Comment (gargnupur) [inline doc/write-a-wasm-extension-with-cpp.md:175]: Not for this PR but for later, I am thinking should this be a part of this repo? We maintain versions of this, similar to Istio versions, import istio e2e test framework too and users write istio integration test using their wasm filter and envoy patches. That way if a breaking change happens like api change in envoy filter or telemetry api, they would know about it immediately?

File: example/plugin.h [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  explicit PluginRootContext(uint32_t id, std::string_view root_id)
+      : RootContext(id, root_id) {}
+
+  bool onConfigure(size_t) override;
+};
+
+class PluginContext : public Context {
+ public:
+  explicit PluginContext(uint32_t id, RootContext* root) : Context(id, root) {}
+
```
Comment (gargnupur) [inline example/plugin.h:14]: should we include all major functions that are available to override here?

File: doc/development-setup.md (hunk 1 above)
Comment (bianpengyuan) [reply doc/development-setup.md:1]: No, we should actually avoid that. It is confusing (Wasm extension development should be independent from the proxy) and not necessary (Envoy development set up is more complex). Oncehttps://github.com/proxy-wasm/proxy-wasm-cpp-sdk/pull/71 is in, we should point to [instruction there](https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/pull/71/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R86) for development set up.

File: doc/write-a-wasm-extension-with-cpp.md (hunk 2 above)
Comment (bianpengyuan) [reply doc/write-a-wasm-extension-with-cpp.md:73]: Somehow go doc does not have istio.io/proxy. Trying to figure out why is it..

File: example/plugin.h (hunk 5 above)
Comment (bianpengyuan) [reply example/plugin.h:14]: No, this is deemed to be super simple. Proxy sdk should document available callbacks, which is current missing.

Comment (bianpengyuan) [reply example/plugin.h:14]: Created https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/78 to track this.

File: doc/write-a-wasm-extension-with-cpp.md (hunk 4 above)
Comment (bianpengyuan) [reply doc/write-a-wasm-extension-with-cpp.md:175]: Yeah what you described is the plan. Integration test is already done for basic_auth filter and should be required for any future extension. I am about to cut 1.8 release branch in this repo, and modules will be released and versioned as Istio.

File: doc/write-a-wasm-extension-with-cpp.md (hunk 2 above)
Comment (bianpengyuan) [reply doc/write-a-wasm-extension-with-cpp.md]: Actually it was there: https://godoc.org/github.com/istio/proxy/test/envoye2e

**Comment by bianpengyuan** [comment]
@mandarjog @kyessenov @gargnupur Any help on reviewing this? thanks!

**Review by gargnupur — APPROVED** [review]

**Pull request merged**